controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -154,113 +125,39 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
-            }
-        };
-
-        WireCommands.TruncateSegment request = new WireCommands.TruncateSegment(requestId, qualifiedName, offset, delegationToken);
-        sendRequestAsync(request, replyProcessor, result, ModelHelper.encode(uri));
-        return result;
+        return connection.sendRequest(requestId, new WireCommands.TruncateSegment(requestId, qualifiedName, offset, delegationToken))
+                .thenApply(r -> transformReply(r, expectedReplies, connection));
     }
 
     public CompletableFuture<Boolean> deleteSegment(final String scope,
                                                     final String stream,
                                                     final long segmentId,
                                                     String delegationToken,
                                                     final long clientRequestId) {
-        final CompletableFuture<Boolean> result = new CompletableFuture<>();
         final Controller.NodeUri uri = getSegmentUri(scope, stream, segmentId);
         final String qualifiedName = getQualifiedStreamSegmentName(scope, stream, segmentId);
-        final WireCommandType type = WireCommandType.DELETE_SEGMENT;
-        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
-
-        final FailingReplyProcessor replyProcessor = new FailingReplyProcessor() {
-
-            @Override
-            public void connectionDropped() {
-                log.warn(requestId, "deleteSegment {} Connection dropped", qualifiedName);
-                result.completeExceptionally(
-                        new WireCommandFailedException(type, WireCommandFailedException.Reason.ConnectionDropped));
-            }
-
-            @Override
-            public void wrongHost(WireCommands.WrongHost wrongHost) {
-                log.warn(requestId, "deleteSegment {} wrong host", qualifiedName);
-                result.completeExceptionally(new WireCommandFailedException(type, WireCommandFailedException.Reason.UnknownHost));
-            }
-
-            @Override
-            public void noSuchSegment(WireCommands.NoSuchSegment noSuchSegment) {
-                log.info(requestId, "deleteSegment {} NoSuchSegment", qualifiedName);
-                result.complete(true);
-            }
-
-            @Override
-            public void segmentDeleted(WireCommands.SegmentDeleted segmentDeleted) {
-                log.info(requestId, "deleteSegment {} SegmentDeleted", qualifiedName);
-                result.complete(true);
-            }
-
-            @Override
-            public void processingFailure(Exception error) {
-                log.error(requestId, "deleteSegment {} failed", qualifiedName, error);
-                handleError(error, result, type);
-            }
+        RawClient connection = new RawClient(ModelHelper.encode(uri), connectionFactory);
+        final long requestId = connection.getFlow().getNextSequenceNumber();
```

Comment:
We can also use a requestId with value `connection.getFlow().asLong()` as we are creating a new RawClient instance. A similar comment can applies to all other methods of SegmentHelper.

Reply:
you mean it's an option or you suggest to use getFlow().asLong() instead getFlow().getNextSequenceNumber()?

Reply:
yes, we can use `getFlow().asLong() ` since we are using the a new RawClient for every request.
`getFlow().getNextSequenceNumber()` is not required since we are not reusing the `RawClient`.